### ISSUE TITLE
String lib definition is not a subtype of indexed collections

### ISSUE THREAD

**Issue by zacharygolba** [facebook/flow#4716]
Currently the [`String` lib def](/facebook/flow/blob/master/lib/core.js#L290) does not implement the following interface _([example](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAdgFwKYBOUAhgMZ5gCSWAJngB540DCcMMepOGcWAPABUAfGAb5aAZyr4CxAEYdBIgN6owYANoBrPAE8AXGCwBXALZzCAXUMCANGrAcsAcxwALQyfOF7AX3RQxlhcPFhgUBgEEjhKABQAbsQwxniG1HSMLGwcIbxKAJQ2YKpgBHg4xgRhicl4GgAMlqj+qKS80WDEnmYWBGAAvOGR0bEaAIy2YABMkwDMlvmt7ThgcobRBNjOA0NROLEA5PjRB4tAA))_:

```javascript
interface IndexedCollection<T> extends Iterable<T> {
  [key: number]: T,
  length: number,
}
```

However, (internally?) strings are treated as such _([example](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAKA5AFwKYDOOWA2gAwC6AXGEQE4CWAdgOYCUqQA))_. A workaround could be achieved with a union but it is lossy in comparison to the `IndexedCollection` interface. You still have to refine `string | T` and a lot of the benefits of parametric polymorphism are nullified.

```javascript
// 😃
function forEach1<T>(fn: T => void, value: IndexedCollection<T>): void {
  for (let i = 0; i < value.length; i++) {
    fn(value[i])
  }
}

// 😐
function forEach2<T>(fn: (string | T) => void, value: string | IndexedCollection<T>): void {
  for (let i = 0; i < value.length; i++) {
    fn(value[i])
  }
}
```

**tldr;**

Is it sound to add a `[key: number]: string` property to the lib def for `String`? If so, I would be happy to create a PR.

**Comment by asolove** [comment]
Huh, it's surprising to me that this doesn't work. I wonder if someone else is aware of some super type that can be used for a function to accept strings or arrays when all it wants to do is subscript them.

**Comment by zacharygolba** [comment]
> Huh, it's surprising to me that this doesn't work.

Likewise!

I submitted a PR that adds the indexer property to `String`. It seems like an easy (yet naive) solution to this problem.

**Comment by asolove** [comment]
Thanks for the PR! I'm really curious to see what the core team says about this. Might be just an oversight or there might be some deep reason why it can't be there.

**Comment by zacharygolba** [comment]
> Thanks for the PR!

Anytime!

> I'm really curious to see what the core team says about this.

Likewise. I also have a feeling that there is more to this issue.
